Add EventWaitHandle creation extension method that takes an ACL

carlossanlop opened the pull request:
Approved API Proposal: https://github.com/dotnet/corefx/issues/41662

### Description
We don't currently have a way to create an EventWaitHandle with a given ACL in .NET Core. We can modify the ACL, but it would be more secure to have the proper ACL on the object from the start.

### Customer impact
Before this change, customers had to create an EventWaitHandle, then set its ACLs. This presents a few problems:

- Potential security hole as event wait handles can be accessed between creation and modification.
- Porting difficulties as there isn't a 1-1 API replacement

This change addresses those problems by adding a new extension method that allows creating an EventWaitHandle and ensuring the provided ACLs are set during creation.

This change is expected to be backported to 3.1.

carlossanlop commented on src/System.Threading.AccessControl/src/Resources/Strings.resx:
```diff
@@ -1,4 +1,63 @@
 ﻿<root>
+  <!-- 
```
These comments were automatically added.

carlossanlop commented on src/System.Threading.AccessControl/src/System.Threading.AccessControl.csproj:
```diff
@@ -1,25 +1,34 @@
     <GeneratePlatformNotSupportedAssemblyMessage Condition="'$(TargetsNetStandard)' == 'true' and '$(TargetsWindows)' != 'true'">SR.PlatformNotSupported_AccessControl</GeneratePlatformNotSupportedAssemblyMessage>
     <IsPartialFacadeAssembly Condition="'$(TargetsNetFx)' == 'true'">true</IsPartialFacadeAssembly>
     <Configurations>net461-Windows_NT-Debug;net461-Windows_NT-Release;netfx-Windows_NT-Debug;netfx-Windows_NT-Release;netstandard2.0-Debug;netstandard2.0-Release;netstandard2.0-Windows_NT-Debug;netstandard2.0-Windows_NT-Release</Configurations>
+    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
```
I added this because of the new unsafe method.
Is this the correct thing to do to allow building unsafe code in this project or should I do something else?

JeremyKuhne replied:
That's right.

carlossanlop commented on src/System.Threading.AccessControl/src/System/Threading/EventWaitHandleAcl.cs:
```diff
@@ -0,0 +1,110 @@
+            try
+            {
+                // This constructor is private
+                // return new EventWaitHandle(handle);
```
This is why my PR is in draft mode.
I am trying to find a way to create an EventWaitHandle passing a SafeWaitHandle to it.

stephentoub replied:
Given the currently exposed surface area, I'm not sure if there's a better way than constructing a new EventWaitHandle, and then doing something like:
```C#
SafeWaitHandle old = event.SafeWaitHandle;
event.SafeWaitHandle = handle;
old.Dispose();
```

carlossanlop replied:
From the proposal, this new method should only accept named objects:

> _We can not support creating unnamed objects without adding new constructors to the types. With names, we can create and then call the open existing methods. If we want to support unnamed I would suggest we add new OpenExisting() overloads that take the handle that our extension methods create._

stephentoub replied:
That's a circular argument. That comment is based on the premise that it can't be done without a new ctor. I don't believe that's the case.

carlossanlop commented on src/System.Threading.AccessControl/src/System/Threading/EventWaitHandleAcl.cs:
```diff
@@ -0,0 +1,110 @@
+
+            if (handle.IsInvalid)
+            {
+                // We call this in netfx. Is it still needed at this point?
```
Will remove this comment and/or the next line depending on the answer to this question.

carlossanlop replied:
Ok thanks. I'll keep the code and remove the comment.

stephentoub commented on src/System.Threading.AccessControl/src/System/Threading/EventWaitHandleAcl.cs:
```diff
@@ -0,0 +1,110 @@
+                    break;
+                default:
+                    throw new ArgumentOutOfRangeException(nameof(mode), SR.Format(SR.ArgumentOutOfRange_Enum));
+            };
```
Nit: this can be simplified to:
```C#
bool isManualReset = mode switch
{
    EventResetMode.ManualReset => true,
    EventResetMode.AutoReset => false,
    _ => throw new ArgumentOutOfRangeException(nameof(mode), SR.Format(SR.ArgumentOutOfRange_Enum));
};
```

carlossanlop commented on src/System.Threading.AccessControl/src/System/Threading/EventWaitHandleAcl.cs:
```diff
@@ -0,0 +1,110 @@
+                    throw Win32Marshal.GetExceptionForWin32Error(errorCode, name);
+                }
+
+                createdNew = errorCode != Interop.Errors.ERROR_ALREADY_EXISTS;
```
I didn't put it in the right place. It should be outside the `if (handle.IsInvalid)`.

carlossanlop commented on src/System.Threading.AccessControl/src/System/Threading/EventWaitHandleAcl.cs:
```diff
@@ -0,0 +1,110 @@
+                // We call this in netfx. Is it still needed at this point?
+                handle.SetHandleAsInvalid();
+
+                int errorCode = Marshal.GetLastWin32Error();
```
> _I'd follow along with what we have in Core more than what is in Framework. It is factored better in Core and we want to reuse the code ultimately_

Yes, I'm starting to notice that. Thanks for pointing it out.

danmosemsft commented on src/System.Threading.AccessControl/src/System/Threading/EventWaitHandleAcl.cs:
```diff
@@ -0,0 +1,117 @@
+            return CreateEventWaitHandle(initialState, isManualReset, name, out createdNew, eventSecurity);
+        }
+
+        private static unsafe EventWaitHandle CreateEventWaitHandle(bool initialState, bool isManualReset, string name, out bool createdNew, EventWaitHandleSecurity security)
```
Do you need a new method? It's just directly called by the line above.

carlossanlop replied:
The approved API did not have the unsafe keyword, that's the main reason. So I wrapped the unsafe code in its own method.
The other reason is because I feel the code is easier to read/understand if I keep the methods smaller.
Would you still prefer me to merge them? Is it ok to modify the approved API method signature?

stephentoub replied:
> The approved API did not have the unsafe keyword, that's the main reason

`unsafe` isn't actually part of the method signature.  The compiler just lets you put it there as part of authoring the method to say that everything in the method can be unsafe (and also in case parameters use pointers).  `unsafe` will as such never be in the approved API shape because it's not actually in the resulting API signature generated by the compiler.

carlossanlop replied:
Good to know, thanks. I'll merge the methods.

danmosemsft commented on src/System.Threading.AccessControl/src/System/Threading/EventWaitHandleAcl.cs:
```diff
@@ -0,0 +1,117 @@
+                if (aggEx != null)
+                {
+                    throw aggEx;
+                }
```
Why wrap in an AggregateException? That usually used to, well, aggregate several exceptions. Would this work?
```c#
                using (handle) EventWaitHandle.OpenExisting(name);
```

carlossanlop replied:
I'm calling this inside the `fixed` section, and the handle returned by `OpenExisting` needs to be returned by the method.

carlossanlop replied:
Also, I don't think I should call `TryOpenExisting(string name, out EventWaitHandle handle)` because we need to bubble up the exception to the user.

JeremyKuhne replied:
Nothing gets returned when you throw. All you need here is:

``` C#
                try
                {
                    return EventWaitHandle.OpenExisting(name);
                }
                finally
                {
                    // Close our handle as the EventWaitHandle will have it's own.
                    handle.Dispose();
                }
```

JeremyKuhne replied:
Which is what @danmosemsft's comment on using does for you. :)

danmosemsft commented on src/System.Threading.AccessControl/src/System/Threading/EventWaitHandleAcl.cs:
```diff
@@ -0,0 +1,117 @@
+
+        private static void ValidateHandle(SafeWaitHandle handle, string name, out bool createdNew)
+        {
+            Debug.Assert(handle != null);
```
This assert does not seem worth having since it is checked (by not throwing) moments later on line 104. In C#, we do not use Debug.Assert to check for null before dereferencing. We use it to check for bad states that otherwise would not be detected until some time later in the program when the problem would be harder to figure out.

carlossanlop replied:
Ok thanks. Removing it.

danmosemsft commented on src/System.Threading.AccessControl/src/System/Threading/EventWaitHandleAcl.cs:
```diff
@@ -0,0 +1,117 @@
+                throw Win32Marshal.GetExceptionForWin32Error(errorCode, name);
+            }
+
+            createdNew = errorCode != Interop.Errors.ERROR_ALREADY_EXISTS;
```
nit, `createdNew = (errorCode != Interop.Errors.ERROR_ALREADY_EXISTS);`

danmosemsft replied:
It's nice to avoid `out` parameters if possible as they don't compose well (and possibly don't generate such good code)_. Although, it might be unclear if it was the return value.

carlossanlop commented on src/System.Threading.AccessControl/tests/ThreadingAclExtensionsTests.cs:
```diff
@@ -45,5 +54,161 @@ public static void ExistenceTest_Unix()
+            var security = GetEventWaitHandleSecurity(WellKnownSidType.BuiltinUsersSid, (EventWaitHandleRights)rights, accessControl);
+            if (accessControl == AccessControlType.Deny)
+            {
+                Assert.Throws<UnauthorizedAccessException>(() =>
```
@JeremyKuhne I don't think this is the expected behavior, but I added the Assert.Throws to illustrate what I'm seeing:
- In netcore, we throw UnauthorizedAccessException when AccessControlType is deny.
- In netfx, we don't throw anything.
I'm comparing the code to try to understand why the behavior doesn't match different frameworks.

stephentoub commented on src/System.Threading.AccessControl/src/System/Threading/EventWaitHandleAcl.cs:
```diff
@@ -0,0 +1,93 @@
+                {
+                    // Close our handle as the EventWaitHandle will have it's own.
+                    handle.Dispose();
+                }
```
I continue to not understand why we're placing the `name != null` restriction.  Why can't this instead be, e.g.
```C#
SafeWaitHandle handle = Interop.Kernel32.CreateEvent(ref secAttrs, isManualReset, initialState, name);
ValidateHandle(handle, name, out createdNew);

var ewh = new EventWaitHandle(initialState, mode);
var old = ewh.SafeWaitHandle;
ewh.SafeWaitHandle = handle;
old.Dispose();
```
?  Does that not work for some reason?

carlossanlop replied:
It does not work.
Unit tests that are passing with my current code, start failing with callstacks like the one below when I add your suggested change.
The exception shows up when attempting to retrieve the access control from the newly created EventWaitHandle.
For example:
```
System.Threading.Tests.ThreadingAclExtensionsTests.EventWaitHandle_Create_SpecificParameters(initialState: True, mode: AutoReset, accessControl: Allow, rights: FullControl) [FAIL]

System.ObjectDisposedException : Safe handle has been closed.

Object name: 'SafeHandle'.

Stack Trace:
/_/src/System.Private.CoreLib/shared/System/Runtime/InteropServices/SafeHandle.cs(156,0): at System.Runtime.InteropServices.SafeHandle.DangerousAddRef(Boolean& success)
/_/src/System.Private.CoreLib/src/System/StubHelpers.cs(1747,0): at System.StubHelpers.StubHelpers.SafeHandleAddRef(SafeHandle pHandle, Boolean& success)
at Interop.Advapi32.GetSecurityInfoByHandle(SafeHandle handle, UInt32 objectType, UInt32 securityInformation, IntPtr& sidOwner, IntPtr& sidGroup, IntPtr& dacl, IntPtr& sacl, IntPtr& securityDescriptor)
D:\corefx\src\System.Security.AccessControl\src\System\Security\AccessControl\Win32.cs(134,0): at System.Security.AccessControl.Win32.GetSecurityInfo(ResourceType resourceType, String name, SafeHandle handle, AccessControlSections accessControlSections, RawSecur
ityDescriptor& resultSd)
D:\corefx\src\System.Security.AccessControl\src\System\Security\AccessControl\NativeObjectSecurity.cs(111,0): at System.Security.AccessControl.NativeObjectSecurity.CreateInternal(ResourceType resourceType, Boolean isContainer, String name, SafeHandle handle, Acc
essControlSections includeSections, Boolean createByName, ExceptionFromErrorCode exceptionFromErrorCode, Object exceptionContext)
D:\corefx\src\System.Security.AccessControl\src\System\Security\AccessControl\NativeObjectSecurity.cs(84,0): at System.Security.AccessControl.NativeObjectSecurity..ctor(Boolean isContainer, ResourceType resourceType, SafeHandle handle, AccessControlSections incl
udeSections, ExceptionFromErrorCode exceptionFromErrorCode, Object exceptionContext)
D:\corefx\src\System.Threading.AccessControl\src\System\Security\AccessControl\EventWaitHandleSecurity.cs(125,0): at System.Security.AccessControl.EventWaitHandleSecurity..ctor(SafeWaitHandle handle, AccessControlSections includeSections)
D:\corefx\src\System.Threading.AccessControl\src\System\Threading\ThreadingAclExtensions.cs(17,0): at System.Threading.ThreadingAclExtensions.GetAccessControl(EventWaitHandle handle)
D:\corefx\src\System.Threading.AccessControl\tests\ThreadingAclExtensionsTests.cs(171,0): at System.Threading.Tests.ThreadingAclExtensionsTests.CreateAndVerifyEventWaitHandle(Boolean initialState, EventResetMode mode, EventWaitHandleSecurity expectedSecurity)
D:\corefx\src\System.Threading.AccessControl\tests\ThreadingAclExtensionsTests.cs(124,0): at System.Threading.Tests.ThreadingAclExtensionsTests.EventWaitHandle_Create_SpecificParameters(Boolean initialState, EventResetMode mode, AccessControlType accessControl,
EventWaitHandleRights rights)
```

stephentoub replied:
> It does not work.

Please share the code you used.

JeremyKuhne replied:
> I continue to not understand why we're placing the name != null restriction.

In theory you could, it would be creating and tossing a global object, of course. I think I'd rather just "fake" it and create and modify if we're going to support it. In theory there shouldn't be risk here as you need the handle to the WaitHandle to do anything with it when it is unnamed. If the security is modified before we give it back it should be good...

@carlossanlop can you please share a more specific repro example or explain why it is failing- afaik, it shouldn't.

carlossanlop replied:
> _Please share the code you used._

> _can you please share a more specific repro example or explain why it is failing- afaik, it shouldn't._

@stephentoub @JeremyKuhne yes of course, gladly. I submitted a new commit with the applied suggestion so the CI hits the failure I mentioned. Please take a look.

stephentoub replied:
> Please take a look.

You didn't remove the `using`... that's why you're getting an ObjectDisposedException.

carlossanlop replied:
Yes, thanks for catching it. Now it works.

carlossanlop commented on src/Common/src/CoreLib/Interop/Windows/Kernel32/Interop.EventWaitHandle.cs:
```diff
@@ -13,16 +14,19 @@ internal static partial class Kernel32
+        [DllImport(Libraries.Kernel32, EntryPoint = "OpenEventW", SetLastError = true, CharSet = CharSet.Unicode)]
         internal static extern SafeWaitHandle OpenEvent(uint desiredAccess, bool inheritHandle, string name);
+
+        [DllImport(Libraries.Kernel32, EntryPoint = "CreateEventExW", SetLastError = true, CharSet = CharSet.Unicode, BestFitMapping = false)]
```
From our conversation earlier, @JeremyKuhne, instead of using CreateEventW, I should use CraeteEventExW, so I could pass EVENT_ALL_ACCESS to desiredAccess. The problem is that we already have a PInvoke that calls it, but the first argument is an IntPtr. I didn't want to touch that version so I added another one that accepts a ref. Suggestions about the name are welcome. Is this the correct thing to do?

JeremyKuhne replied:
Don't create another Import as that is unnecessary overhead (additional IL generation for the new import). You can just create a wrapper here that takes the address of the `SECURITY_ATTRIBUTES` and casts to `IntPtr` for convenience.

carlossanlop replied:
@JeremyKuhne 

Both of these PInvokes are hitting the same failure of the object disposed exception I mentioned in my other comment:

```cs
[DllImport(Libraries.Kernel32, EntryPoint = "CreateEventW", SetLastError = true, CharSet = CharSet.Unicode, BestFitMapping = false)]
internal static extern SafeWaitHandle CreateEvent(ref SECURITY_ATTRIBUTES lpSecurityAttributes, bool isManualReset, bool initialState, string? name);
```

```cs
[DllImport(Libraries.Kernel32, EntryPoint = "CreateEventExW", SetLastError = true, CharSet = CharSet.Unicode, BestFitMapping = false)]
 internal static extern SafeWaitHandle CreateEvent(ref SECURITY_ATTRIBUTES lpSecurityAttributes, string? name, uint flags, uint desiredAccess);
```
This second one I was calling it by passing:

```cs
            uint eventFlags = initialState ? Interop.Kernel32.CREATE_EVENT_INITIAL_SET : 0;

            if (mode == EventResetMode.ManualReset)
            {
                EventResetMode.ManualReset => true,
                EventResetMode.AutoReset => false,
                _ => throw new ArgumentOutOfRangeException(nameof(mode), SR.Format(SR.ArgumentOutOfRange_Enum))
            };
                eventFlags |= (uint)Interop.Kernel32.CREATE_EVENT_MANUAL_RESET;
            }
            ...
            using SafeWaitHandle handle = Interop.Kernel32.CreateEvent(ref secAttrs, name, eventFlags, (uint)EventWaitHandleRights.FullControl);

```

carlossanlop replied:
Nevermind! Stephen pointed out that I should not be using the using keyword anymore.

Should I still switch to use the second PInvoke?

JeremyKuhne commented on src/System.Threading.AccessControl/src/System/Threading/EventWaitHandleAcl.cs:
```diff
@@ -0,0 +1,86 @@
+            {
+                throw new ArgumentException(SR.Format(SR.Argument_WaitHandleNameTooLong, name), nameof(name));
+            }
+            if (eventSecurity == null)
```
nit: no line break after block

JeremyKuhne commented on src/Common/src/CoreLib/Interop/Windows/Kernel32/Interop.EventWaitHandle.cs:
```diff
@@ -13,16 +14,19 @@ internal static partial class Kernel32
+        [DllImport(Libraries.Kernel32, EntryPoint = "OpenEventW", SetLastError = true, CharSet = CharSet.Unicode)]
         internal static extern SafeWaitHandle OpenEvent(uint desiredAccess, bool inheritHandle, string name);
+
+        [DllImport(Libraries.Kernel32, EntryPoint = "CreateEventW", SetLastError = true, CharSet = CharSet.Unicode, BestFitMapping = false)]
```
Please don't create another import, get the address yourself, either here in a helper or in the caller to the other import.

carlossanlop replied:
Removing.

JeremyKuhne commented on src/System.Threading.AccessControl/src/System/Threading/EventWaitHandleAcl.cs:
```diff
@@ -0,0 +1,86 @@
+
+                ValidateHandle(handle, name, out createdNew);
+
+                var ewh = new EventWaitHandle(initialState, mode);
```
I don't like creating global objects when they aren't needed. Can we condition this on their being a name?

Additionally, as I stated, I'd rather modify the existing security when there is a name, for the same reason- I think it's better not to create throwaway global objects if possible.

stephentoub replied:
@JeremyKuhne, can you help me understand better why?  It's unnamed and very short-lived.  It's also faster:
```C#
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Diagnosers;
using BenchmarkDotNet.Running;
using System;
using System.Threading;

[MemoryDiagnoser]
public class Program
{
    static void Main(string[] args) => BenchmarkSwitcher.FromAssemblies(new[] { typeof(Program).Assembly }).Run(args);

    private static readonly string s_name = Guid.NewGuid().ToString("N");
    private EventWaitHandle _named = new EventWaitHandle(false, EventResetMode.ManualReset, s_name);

    [Benchmark]
    public void CreateNew() => new EventWaitHandle(false, EventResetMode.ManualReset).Dispose();

    [Benchmark]
    public void OpenExisting() => EventWaitHandle.OpenExisting(s_name).Dispose();
}
```

|       Method |     Mean |     Error |    StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------- |---------:|----------:|----------:|-------:|------:|------:|----------:|
|    CreateNew | 1.346 us | 0.0041 us | 0.0034 us | 0.0076 |     - |     - |      56 B |
| OpenExisting | 2.121 us | 0.0069 us | 0.0057 us | 0.0076 |     - |     - |      56 B |

JeremyKuhne replied:
My primary concern is that we're forcing more contention in creating twice as many synchronizing objects. To be fair, I haven't spent much time looking at the Windows code here, so it is more of a gut concern. If you feel comfortable with it, I'm ok with trying the current route.

JeremyKuhne commented on src/System.Threading.AccessControl/src/System/Threading/EventWaitHandleAcl.cs:
```diff
@@ -0,0 +1,86 @@
+                throw Win32Marshal.GetExceptionForWin32Error(errorCode, name);
+            }
+
+            createdNew = (errorCode != Interop.Errors.ERROR_ALREADY_EXISTS);
```
Do you have a test case for `createdNew`?

carlossanlop replied:
No, I don't. I should add one.

carlossanlop replied:
Added.

stephentoub commented on src/System.Threading.AccessControl/src/System/Threading/EventWaitHandleAcl.net46.cs:
```diff
@@ -0,0 +1,38 @@
+            out bool createdNew,
+            EventWaitHandleSecurity eventSecurity)
+        {
+            if (string.IsNullOrEmpty(name))
```
name should be allowed to be null

carlossanlop replied:
Fixing.

stephentoub commented on src/System.Threading.AccessControl/src/System/Threading/EventWaitHandleAcl.cs:
```diff
@@ -0,0 +1,86 @@
+        /// An <see cref="EventWaitHandle" /> with system-wide name <paramref name="name" /> cannot be created. An <see cref="EventWaitHandle" /> of a different type might have the same name.</exception>
+        public static unsafe EventWaitHandle Create(bool initialState, EventResetMode mode, string name, out bool createdNew, EventWaitHandleSecurity eventSecurity)
+        {
+            if (name.Length > Interop.Kernel32.MAX_PATH)
```
name should be allowed to be null

carlossanlop replied:
Fixing.

JeremyKuhne replied:
Tbh, I'd rather not check for the length at all here. Have we tested to see that it won't work without this? Like, is it `MAX_PATH` + prefix? (e.g. `GLOBAL\ReallyLongName`) I'd prefer to just let Windows report a failure and not inadvertently block some scenario.

carlossanlop replied:
Sure, that's another option. I'll add a unit test and see what exception is thrown by Windows, if any.

carlossanlop replied:
Update: Windows does not seem to be throwing an exception if we choose a name that goes beyond MAX_PATH.

carlossanlop replied:
@JeremyKuhne  We do throw in Framework. Should we match the behavior in .NET Core?
https://referencesource.microsoft.com/#mscorlib/system/threading/eventwaithandle.cs,59

carlossanlop replied:
I'm probably going to answer my own question and say no, because we don't throw on MAX_LENGTH in the .NET Core constructor for EventWaitHandle:
https://source.dot.net/#System.Private.CoreLib/shared/System/Threading/EventWaitHandle.Windows.cs,20

JeremyKuhne commented on src/System.Threading.AccessControl/src/System/Threading/EventWaitHandleAcl.cs:
```diff
@@ -0,0 +1,88 @@
+                    lpSecurityDescriptor = (IntPtr)pSecurityDescriptor
+                };
+
+                SafeWaitHandle handle = Interop.Kernel32.CreateEventEx((IntPtr)(&secAttrs), name, eventFlags, (uint)EventWaitHandleRights.FullControl);
```
nit: would you mind breaking the arguments into separate lines? It is getting a little long.

carlossanlop replied:
Sure no problem. Let me accumulate a few changes to avoid sending too many commits :)

danmosemsft commented:
BTW, unlike System.IO.FileSystem.AccessControl,  System.Threading.AccessControl package is not in netcoreapp. Therefore if this was backported to servicing (which I was assuming would not happen anyway, since today's the last day), it would presumably need package authoring as well .cc @ericstj

JeremyKuhne commented on src/System.Threading.AccessControl/src/System/Threading/EventWaitHandleAcl.net46.cs:
```diff
@@ -0,0 +1,31 @@
+            out bool createdNew,
+            EventWaitHandleSecurity eventSecurity)
+        {
+            if (eventSecurity == null)
```
I wouldn't do any validation here that the method we call is going to do anyway. And in this case, you're allowed to pass null for security, it just assumes default. I would follow suit with the constructor for .NET- allow the null.

carlossanlop replied:
Makes sense. Removing.

carlossanlop commented on src/System.Threading.AccessControl/tests/ThreadingAclExtensionsTests.cs:
```diff
@@ -45,5 +53,247 @@ public static void ExistenceTest_Unix()
+
+        [Fact]
+        [PlatformSpecific(TestPlatforms.Windows)]
+        public static void EventWaitHandle_Create_EmptyNameMultipleNew()
```
This matches the behavior of the existing `EventWaitHandle` constructors that take no security object: Two nameless EWH events created in a row will both be created new.

carlossanlop replied:
@JeremyKuhne told me this may not be the correct behavior. I'm investigating.

carlossanlop replied:
Nevermind, this is correct.

carlossanlop commented on src/System.Threading.AccessControl/tests/ThreadingAclExtensionsTests.cs:
```diff
@@ -45,5 +53,247 @@ public static void ExistenceTest_Unix()
+
+        [Fact]
+        [PlatformSpecific(TestPlatforms.Windows)]
+        public static void EventWaitHandle_Create_CreateNewExisting()
```
This matches the behavior of the existing `EventWaitHandle` constructors that take no security object: Two named EWH events created in a row, will only have the first one created new, and the second one will be opened.

JeremyKuhne commented on src/System.Threading.AccessControl/src/System/Threading/EventWaitHandleAcl.cs:
```diff
@@ -0,0 +1,88 @@
+        /// An <see cref="EventWaitHandle" /> with system-wide name <paramref name="name" /> cannot be created. An <see cref="EventWaitHandle" /> of a different type might have the same name.</exception>
+        public static unsafe EventWaitHandle Create(bool initialState, EventResetMode mode, string name, out bool createdNew, EventWaitHandleSecurity eventSecurity)
+        {
+            if (eventSecurity == null)
```
I wouldn't check for this as the method we're trying to support (the Framework constructor) allows it- it just doesn't set the descriptor when it is null.

carlossanlop replied:
Correct, done. I updated the code here so that in the case the eventSecurity is null, we directly return an EventWaitHandle instance by calling the old constructor.

ericstj commented:
> package is not in netcoreapp. Therefore if this was backported to servicing 

Actually all of these have packages, so they all need care in servicing.  The distinction is not "is in netcoreapp" it is "does it have a package".  If a library has a package, we need include that package in the build to ship it in servicing.

carlossanlop commented:
@danmosemsft @ericstj 
Do we still want to port this to 3.1?

danmosemsft commented:
Throw up a PR and template when this is in, but it's likely to late.

carlossanlop commented on src/System.Threading.AccessControl/src/System/Threading/EventWaitHandleAcl.cs:
```diff
@@ -0,0 +1,90 @@
+{
+    public static class EventWaitHandleAcl
+    {
+        /// <summary>Creates a new <see cref="EventWaitHandle" /> instance, ensuring it is created with the specified event security.</summary>
```
```suggestion
        /// <summary>Gets or creates an <see cref="EventWaitHandle" /> instance, allowing a <see cref="EventWaitHandleSecurity " /> instance to be optionally specified to set it during the event creation.</summary>
```

carlossanlop commented on src/System.Threading.AccessControl/src/System/Threading/EventWaitHandleAcl.cs:
```diff
@@ -0,0 +1,90 @@
+        /// <exception cref="DirectoryNotFoundException">Could not find a part of the path specified in <paramref name="name" />.</exception>
+        /// <exception cref="WaitHandleCannotBeOpenedException">A system-wide synchronization event with the provided <paramref name="name" /> was not found.
+        /// -or-
+        /// An <see cref="EventWaitHandle" /> with system-wide name <paramref name="name" /> cannot be created. An <see cref="EventWaitHandle" /> of a different type might have the same name.</exception>
```
```suggestion
        /// An <see cref="EventWaitHandle" /> with system-wide name <paramref name="name" /> cannot be created. An <see cref="EventWaitHandle" /> of a different type might have the same name.</exception>
        <remarks>If a `name` is passed and the system event already exists, the existing event is returned. If `name` is `null` or <see cref="string.Empty" />, a new local event is always created.</remarks>
```

carlossanlop commented on src/System.Threading.AccessControl/src/System/Threading/EventWaitHandleAcl.cs:
```diff
@@ -0,0 +1,88 @@
+        /// <param name="mode">One of the enum values that determines whether the event resets automatically or manually.</param>
+        /// <param name="name">The name, if the event is a system-wide synchronization event; otherwise, <see langword="null" /> or an empty string.</param>
+        /// <param name="createdNew">When this method returns, it is set to <see langword="true" /> if a local event was created (that is, if name is <see langword="null" /> or an empty string) or if the specified named system event was created; <see langword="false" /> if the specified named system event already existed. This parameter is passed uninitialized.</param>
+        /// <param name="eventSecurity">The Windows access control security to apply.</param>
```
```suggestion
        /// <param name="eventSecurity">The optional Windows access control security to apply.</param>
```

carlossanlop commented on src/System.Threading.AccessControl/src/System/Threading/EventWaitHandleAcl.cs:
```diff
@@ -0,0 +1,88 @@
+        /// <param name="initialState"><see langword="true" /> to set the initial state to signaled if the named event is created as a result of this call; <see langword="false" /> to set it to non-signaled.</param>
+        /// <param name="mode">One of the enum values that determines whether the event resets automatically or manually.</param>
+        /// <param name="name">The name, if the event is a system-wide synchronization event; otherwise, <see langword="null" /> or an empty string.</param>
+        /// <param name="createdNew">When this method returns, it is set to <see langword="true" /> if a local event was created (that is, if name is <see langword="null" /> or an empty string) or if the specified named system event was created; <see langword="false" /> if the specified named system event already existed. This parameter is passed uninitialized.</param>
```
```suggestion
        /// <param name="createdNew">When this method returns, this argument is always set to <see langword="true" /> if a local event is created; that is, when <paramref name="name" /> is <see langword="null" /> or <see cref="string.Empty" />. If <paramref name="name" /> has a valid, non-empty value, this argument is set to <see langword="true" /> when the system event is created, or it is set to <see langword="false" /> if an existing system event is found with that name. This parameter is passed uninitialized.</param>
```

carlossanlop commented:
@danmosemsft @stephentoub @JeremyKuhne the unit tests passed and I have addressed all comments. Would you mind taking another look and letting me know if I can merge?

danmosemsft commented:
I don't have any more feedback but I'll let one of the others do the sign off.

carlossanlop commented on src/System.Threading.AccessControl/src/System/Threading/EventWaitHandleAcl.cs:
```diff
@@ -0,0 +1,88 @@
+
+                ValidateHandle(handle, name, out createdNew);
+
+                var ewh = new EventWaitHandle(initialState, mode);
```
```suggestion
                EventWaitHandle ewh = new EventWaitHandle(initialState, mode);
```

carlossanlop commented on src/System.Threading.AccessControl/src/System/Threading/EventWaitHandleAcl.cs:
```diff
@@ -0,0 +1,88 @@
+                ValidateHandle(handle, name, out createdNew);
+
+                var ewh = new EventWaitHandle(initialState, mode);
+                var old = ewh.SafeWaitHandle;
```
```suggestion
                SafeWaitHandle old = ewh.SafeWaitHandle;
```

maryamariyan commented:
Thank you for your contribution. As announced in dotnet/coreclr#27549 this repository will be moving to dotnet/runtime on November 13. If you would like to continue working on this PR after this date, the easiest way to move the change to dotnet/runtime is:
1.	In your corefx repository clone, create patch by running `git format-patch origin`
2.	In your runtime repository clone, apply the patch by running `git apply --directory src/corefx <path to the patch created in step 1>`

stephentoub commented on src/System.Threading.AccessControl/tests/EventWaitHandleAclTests.cs:
```diff
@@ -0,0 +1,269 @@
+        [InlineData(false, EventResetMode.ManualReset, EventWaitHandleRights.Modify,      AccessControlType.Allow)]
+        [InlineData(false, EventResetMode.ManualReset, EventWaitHandleRights.Modify,      AccessControlType.Deny)]
+        [InlineData(false, EventResetMode.ManualReset, EventWaitHandleRights.Modify | EventWaitHandleRights.Synchronize, AccessControlType.Allow)]
+        [InlineData(false, EventResetMode.ManualReset, EventWaitHandleRights.Modify | EventWaitHandleRights.Synchronize, AccessControlType.Deny)]
```
If you're trying to create a matrix of inputs, it'd be cleaner and easier to understand if you use a MemberData, e.g.
```C#
public static IEnumerable<object[]> EventWaitHandle_Create_SpecificParameters_MemberData() =>
    from initialState in new[] { true, false }
    from mode in new[] { EventResetMode.AutoReset, EventResetMode.ManualReset }
    from rights in new[] { EventWaitHandleRights.FullControl, EventWaitHandleRights.Synchronize, EventWaitHandleRights.Modify, EventWaitHandleRights.Modify | EventWaitHandleRights.Synchronize }
    from accessControl in new[] { AccessControlType.Allow, AccessControlType.Deny }
    select new object[] { initialState, mode, rights accessControl };

[Theory]
[MemberData(nameof(EventWaitHandle_Create_SpecificParameters_MemberData))]
public void EventWaitHandle_Create_SpecificParameters(bool initialState, EventResetMode mode, EventWaitHandleRights rights, AccessControlType accessControl)
{
    ...
}
```

carlossanlop replied:
Nice! Didn't know about that option. Thanks for letting me know. Fixing it.

danmosemsft replied:
That is a neat approach!

stephentoub commented on src/System.Threading.AccessControl/tests/EventWaitHandleAclTests.cs:
```diff
@@ -0,0 +1,269 @@
+        {
+            Assert.Throws<DirectoryNotFoundException>(() =>
+            {
+                string prefixedName = @"GLOBAL\" + GetRandomName();
```
Nit: move this out of the Assert.Throws.  In general you want the Assert.Throws to be as tight around the thing that's going to throw as possible.

carlossanlop replied:
Makes sense. Fixing.

stephentoub commented on src/System.Threading.AccessControl/tests/EventWaitHandleAclTests.cs:
```diff
@@ -0,0 +1,269 @@
+        public void EventWaitHandle_Create_BeyondMaxPathLength()
+        {
+            string name = new string('x', Interop.Kernel32.MAX_PATH + 100);
+            var security = GetBasicEventWaitHandleSecurity();
```
Nit: var => whatever type this is

stephentoub commented on src/System.Threading.AccessControl/tests/EventWaitHandleAclTests.cs:
```diff
@@ -0,0 +1,269 @@
+        [InlineData("")]
+        public void EventWaitHandle_Create_NameMultipleNew(string name)
+        {
+            var security = GetBasicEventWaitHandleSecurity();
```
Nit: var => whatever type this is

stephentoub commented on src/System.Threading.AccessControl/tests/EventWaitHandleAclTests.cs:
```diff
@@ -0,0 +1,269 @@
+        public void EventWaitHandle_Create_CreateNewExisting()
+        {
+            string name = GetRandomName();
+            var security = GetBasicEventWaitHandleSecurity();
```
Ditto
